pandas/core/groupby/grouper.py

```diff
@@ -663,9 +683,26 @@ def _codes_and_uniques(self) -> tuple[npt.NDArray[np.signedinteger], ArrayLike]:
             uniques = Categorical.from_codes(
                 codes=ucodes, categories=categories, ordered=cat.ordered
             )
+            if not self._dropna and np.any(cat.codes < 0):
```

Comment:
Nit: Could save a `cat.codes < 0` since it looks used at least one more time in the if/else branch

Reply:
Thanks - reworked this section with your observation.